README.md
# Minter JS Wallet

[![NPM Package](https://img.shields.io/npm/v/minterjs-wallet.svg?style=flat-square)](https://www.npmjs.org/package/minterjs-wallet)
[![Build Status](https://img.shields.io/travis/com/MinterTeam/minterjs-wallet/master.svg?style=flat-square)](https://travis-ci.com/MinterTeam/minterjs-wallet)
[![Coverage Status](https://img.shields.io/coveralls/github/MinterTeam/minterjs-wallet/master.svg?style=flat-square)](https://coveralls.io/github/MinterTeam/minterjs-wallet?branch=master)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg?style=flat-square)](https://github.com/MinterTeam/minterjs-wallet/blob/master/LICENSE)

A lightweight Minter Wallet implementation.
Forked from [ethereumjs-wallet](https://github.com/ethereumjs/ethereumjs-wallet)

It is complemented by the following packages:
- [minter-js-sdk](https://github.com/MinterTeam/minter-js-sdk) complete JS solution to work with Minter
- [minterjs-tx](https://github.com/MinterTeam/minterjs-tx) to sign transactions
- [minterjs-util](https://github.com/MinterTeam/minterjs-util) utility functions

Motivations are:
- be lightweight
- work in a browser
- support BIP39 mnemonic phrases
- support BIP32 HD keys

Features not supported:
- signing transactions
- managing storage (neither in node.js or the browser)
- conversion between various wallet formats

## Install

```npm install minterjs-wallet```

or from browser

```html
<script src="https://unpkg.com/minterjs-wallet"></script>
<script>
const wallet = minterWallet.generateWallet();
const wallet2 = minterWallet.walletFromMnemonic('...');
// use async for better performance
const wallet3 = await minterWallet.walletFromMnemonicAsync('...');
</script>
```

## Wallet API

### Construction methods:

#### `generateWallet()`
Generates wallet from a random BIP39 mnemonic phrase (uses `bip39.generateMnemonic()` under the hood).
```js
import {generateWallet} from 'minterjs-wallet';
const wallet = generateWallet();
```

#### `walletFromMnemonic(mnemonic)`
Create a wallet instance based on BIP39 12 words mnemonic phrase
```js
import {walletFromMnemonic} from 'minterjs-wallet';
const wallet = walletFromMnemonic('surround tape away million into program organ tonight write prefer inform cool');
```

#### `walletFromMnemonicAsync(mnemonic)`
Same as `walletFromMnemonic` but async and has better performance in browser, beacause it uses `window.crypto.subtle` under hood
```js
import {walletFromMnemonicAsync} from 'minterjs-wallet';
const wallet = await walletFromMnemonicAsync('surround tape away million into program organ tonight write prefer inform cool');
```

#### `walletFromPrivateKey(privateKey)`
Create a wallet instance based on a raw private key
```js
import {walletFromPrivateKey} from 'minterjs-wallet';
const privateKeyBuffer = Buffer.from('ef2af2385681c490bc473c2f7e6097fc1a38e6f67b44b81a3350a6583aadd144', 'hex')
const wallet = walletFromPrivateKey(privateKeyBuffer);
```


### Wallet instance methods:

#### `.getAddress()`
Return the address: 20 bytes length Buffer or Uint8Array
```js
wallet.getAddress();
// [88,108,62,182,16,84,7,36,101,31,244,29,32,158,20,41,238,62,68,154]
```

#### `.getAddressString()`
Return the Minter-style address
```js
wallet.getAddress();
// 'Mx586c3eb610540724651ff41d209e1429ee3e449a'
```

#### `.getMnemonic()`
Return the mnemonic phrase: 12 words string.
**Note:** Only works with instance created/generated from mnemonic, otherwise it will throw an error.
```js
wallet.getMnemonic();
// 'surround tape away million into program organ tonight write prefer inform cool'
```

#### `.getPrivateKey()`
Return the private key: 32 bytes length Buffer or Uint8Array
```js
wallet.getPrivateKey();
// [239,42,242,56,86,129,196,144,188,71,60,47,126,96,151,252,26,56,230,246,123,68,184,26,51,80,166,88,58,173,209,68]
```

#### `.getPrivateKeyString()`
Return the private key string of 64 hex characters
```js
wallet.getPrivateKeyString();
// 'ef2af2385681c490bc473c2f7e6097fc1a38e6f67b44b81a3350a6583aadd144'
```

#### `.getPublicKey()`
Return the uncompressed Ethereum-style public key: 64 bytes length Buffer or Uint8Array
```js
wallet.getPublicKey();
// [251,82,201,189,133,251,174,27,6,6,18,34,12,222,116,254,99,169,65,249,135,81,170,13,35,99,50,6,231,95,48,69,41,47,96,75,240,242,9,77,23,168,173,59,137,223,128,80,144,69,34,91,145,21,255,133,112,189,68,8,42,245,210,116]
```

#### `.getPublicKeyString()`
Return the Minter-style public key string
```js
wallet.getPublicKeyString();
// 'Mpfb52c9bd85fbae1b060612220cde74fe63a941f98751aa0d23633206e75f3045'
```

### Mnemonic

#### `generateMnemonic()`
Generate random 12 words mnemonic phrase. Exposed `bip39.generateMnemonic()`.
```js
import {generateMnemonic} from 'minterjs-wallet';
const mnemonic = generateMnemonic();
// 'surround tape away million into program organ tonight write prefer inform cool'
```

#### `isValidMnemonic(mnemonic)`
Check that given mnemonic is valid, returns `boolean`, uses `bip39.validateMnemonic()` under the hood.
```js
import {isValidMnemonic} from 'minterjs-wallet';
const isValid = isValidMnemonic('surround tape away million into program organ tonight write prefer inform cool');
// true
```

## License

MIT License


CHANGELOG.md
## 6.1.0 - 2021-03-24
- add `walletFromMnemonicAsync`, use it to get better performance in browser

## 6.0.0 - 2020-08-11
- **BREAKING** getPrivateKeyString now returns 0x-prefixed string

## 5.0.0 - 2020-08-11
- **BREAKING** remove walletFromExtendedPrivateKey
- update deps

## 4.0.4 - 2019-11-13
- drop safe-buffer dep
- drop des.js dep, it's updated and no need to dedupe more

## 4.0.3 - 2019-11-13
- update deps

## 4.0.2 - 2019-10-01
- update deps

## 4.0.1 - 2019-08-15
- update deps

## 4.0.0 - 2019-07-30
- **BREAKING** rename UMD module from `minterJsTx` to `minterWallet`
- fix browser usage of UMD module
- tweak bundle tests

## 3.0.4 - 2019-07-08
- update deps

## 3.0.3 - 2019-06-05
- update deps

## 3.0.2 - 2019-05-14
- update deps

## 3.0.1 - 2019-02-28
- fix package.json "browser" field

## 3.0.0 - 2019-02-27
Align package structure to other minterjs packages.
- **BREAKING** dist/index.js now contain UMD build without es modules
- add UMD and commonjs builds
- replace `mocha` with `jest`
- replace `standard` with `eslint`'s airbnb config
- update deps

## 2.2.2 - 2019-02-18
- Update deps
- fix ethereumjs-util

## 2.2.1 - 2018-11-30
- Update deps

## 2.2.0 - 2018-11-97
- Add `generateMnemonic()` and `isValidMnemonic()` methods

## 2.1.1 - 2018-08-09
- Move source to /src directory (from ethereumjs-wallet merge)
- Add `module` field to package.json

## 2.1.0 - 2018-07-18
- Change public key string format to Minter style `Mp...`
- Explicitly set options to `babel-register` instead of `.babelrc` lookup

## 2.0.4 - 2018-07-12
- Separate construction methods from Wallet instance

## 2.0.1 - 2018-07-10
- Add generation from random BIP39 mnemonic

## 2.0.0 - 2018-07-10
- Added creation from BIP39 mnemonic
- Removed unused Ethereum methods from Wallet
- Removed HDKey wrapper, use [hdkey](https://github.com/cryptocoinjs/hdkey) instead
- Removed thirdparty generation
- Removed provider engine
- Updated to ES2015 syntax

## 1.0.1 - 2018-05-29
- fork from [ethereumjs-wallet](https://github.com/ethereumjs/ethereumjs-wallet)
- update address string format to `Mx`
- remove `0x` from keys string format


## [0.6.1] - 2018-07-28
- Added support for vanity address generation, PR [#5](https://github.com/ethereumjs/ethereumjs-wallet/pull/5)
- Fixed typo in provider-engine, PR [#16](https://github.com/ethereumjs/ethereumjs-wallet/pull/16)
- Accept the true range of addresses for ICAP direct, PR [#6](https://github.com/ethereumjs/ethereumjs-wallet/pull/6)
- Switched to babel ES5 build, PR [#37](https://github.com/ethereumjs/ethereumjs-wallet/pull/37)
- Improve test coverage (at 88% now), PR [#27](https://github.com/ethereumjs/ethereumjs-wallet/pull/27)
- Various dependency updates, PR [#25](https://github.com/ethereumjs/ethereumjs-wallet/pull/25)

[0.6.1]: https://github.com/ethereumjs/ethereumjs-wallet/compare/v0.6.0...v0.6.1

## [0.6.0] - 2016-04-27
- Added provider-engine integration, PR [#7](https://github.com/ethereumjs/ethereumjs-wallet/pull/7)

[0.6.0]: https://github.com/ethereumjs/ethereumjs-wallet/compare/v0.5.2...v0.6.0

## [0.5.2] - 2016-04-25
- Dependency updates

[0.5.2]: https://github.com/ethereumjs/ethereumjs-wallet/compare/v0.5.1...v0.5.2

## [0.5.1] - 2016-03-26
- Bugfix for ``EthereumHDKey.privateExtendedKey()``
- Added travis and coveralls support
- Documentation and test improvements

[0.5.1]: https://github.com/ethereumjs/ethereumjs-wallet/compare/v0.5.0...v0.5.1

## [0.5.0] - 2016-03-23
- Support HD keys using ``cryptocoinjs/hdkey``
- Ensure private keys are valid according to the curve
- Support instantation with public keys
- Support importing BIP32 xpub/xpriv
- Only support Ethereum keys internally, non-strict mode for importing compressed ones
- Thirdparty API doc improvements

[0.5.0]: https://github.com/ethereumjs/ethereumjs-wallet/compare/v0.4.0...v0.5.0

## Older releases:

- [0.4.0](https://github.com/ethereumjs/ethereumjs-wallet/compare/v0.3.0...v0.4.0) - 2016-03-16
- [0.3.0](https://github.com/ethereumjs/ethereumjs-wallet/compare/v0.2.1...v0.3.0) - 2016-03-09
- [0.2.1](https://github.com/ethereumjs/ethereumjs-wallet/compare/v0.2.0...v0.2.1) - 2016-03-07
- [0.2.0](https://github.com/ethereumjs/ethereumjs-wallet/compare/v0.1.0...v0.2.0) - 2016-03-07
- 0.1.0 - 2016-02-23


test/index.test.js
import assert from 'assert';
import * as bip39 from 'bip39';
import { isValidPrivate } from 'ethereumjs-util';
import Wallet, {
    seedFromMnemonic, seedFromMnemonicAsync, hdKeyFromSeed, walletFromPrivateKey, walletFromMnemonic, walletFromMnemonicAsync, generateWallet, isValidMnemonic,
} from '~/src';

const MNEMONIC = 'exercise fantasy smooth enough arrive steak demise donkey true employ jealous decide blossom bind someone';
const PRIVATE_KEY = '0x5fa3a8b186f6cc2d748ee2d8c0eb7a905a7b73de0f2c34c5e7857c3b46f187da';
const fixturewallet = walletFromMnemonic(MNEMONIC);

describe('seedFromMnemonic()', () => {
    const seedBuffer = seedFromMnemonic(MNEMONIC);
    it('should have correct length', () => {
        assert.strictEqual(seedBuffer.length, 64);
    });
    it('should have correct value', () => {
        assert.strictEqual(seedBuffer.toString('hex'), '4ca64bd07f7765099f856e554504900953f59bd874009b4905fca674158ea0563bd7c35b79bff3feb3b33fcbdcdb771b93338ea4f29fcb9cfbb2f570093878c1');
    });
});

describe('seedFromMnemonicAsync()', () => {
    it('should have correct value', async () => {
        const seedBuffer = seedFromMnemonic(MNEMONIC);
        const seedBufferAsync = await seedFromMnemonicAsync(MNEMONIC);
        assert.strictEqual(seedBuffer.toString('hex'), seedBufferAsync.toString('hex'));
    });
});

describe('hd key from mnemonic', () => {
    it('should work', () => {
        const seed = seedFromMnemonic(MNEMONIC);
        const hdKey = hdKeyFromSeed(seed);
        const privateKey = hdKey._privateKey.toString('hex');
        assert.strictEqual(privateKey.length, 64);
        assert.strictEqual(`0x${privateKey}`, PRIVATE_KEY);
    });
});

describe('.getPrivateKey()', () => {
    it('should work', () => {
        assert.strictEqual(`0x${fixturewallet.getPrivateKey().toString('hex')}`, PRIVATE_KEY);
    });
});

describe('.getPrivateKeyString()', () => {
    it('should work', () => {
        assert.strictEqual(fixturewallet.getPrivateKeyString(), PRIVATE_KEY);
    });
});

describe('.getPublicKey()', () => {
    it('should work', () => {
        assert.strictEqual(fixturewallet.getPublicKey().toString('hex'), 'f9e036839a29f7fba2d5394bd489eda927ccb95acc99e506e688e4888082b3a3cb8a015b8031d02e79456aedb361fa20ec1a119d6009e5c08e9d1eeb5b29ad92');
    });
});

describe('.getPublicKeyString()', () => {
    it('should work', () => {
        assert.strictEqual(fixturewallet.getPublicKeyString(), 'Mpf9e036839a29f7fba2d5394bd489eda927ccb95acc99e506e688e4888082b3a3');
    });
});

describe('.getAddress()', () => {
    it('should work', () => {
        assert.strictEqual(fixturewallet.getAddress().toString('hex'), '7633980c000139dd3bd24a3f54e06474fa941e16');
    });
});

describe('.getAddressString()', () => {
    it('should work', () => {
        assert.strictEqual(fixturewallet.getAddressString(), 'Mx7633980c000139dd3bd24a3f54e06474fa941e16');
    });
});

describe('generate random wallet', () => {
    const wallet = generateWallet();
    it('should have valid mnemonic', () => {
        assert.ok(bip39.validateMnemonic(wallet.getMnemonic()));
    });
    it('should have valid private key', () => {
        assert.ok(isValidPrivate(wallet.getPrivateKey()));
    });
});

describe('private key only wallet', () => {
    const privKey = Buffer.from(PRIVATE_KEY.substr(2), 'hex');
    const wallet = walletFromPrivateKey(privKey);
    it('walletFromPrivateKey() should work', () => {
        assert.strictEqual(wallet.getPublicKey().toString('hex'),
            'f9e036839a29f7fba2d5394bd489eda927ccb95acc99e506e688e4888082b3a3cb8a015b8031d02e79456aedb361fa20ec1a119d6009e5c08e9d1eeb5b29ad92');
    });
    it('walletFromPrivateKey() should not accept invalid key', () => {
        assert.throws(() => {
            walletFromPrivateKey(Buffer.from('001122', 'hex'));
        });
    });
    it('.getAddress() should work', () => {
        assert.strictEqual(wallet.getAddress().toString('hex'), '7633980c000139dd3bd24a3f54e06474fa941e16');
    });
    it('.getMnemonic() should fail', () => {
        assert.throws(() => {
            wallet.getMnemonic();
        }, /^Error: This is a private key only wallet$/);
    });
});

describe('Wallet', () => {
    it('constructor should throw with 2 arguments', () => {
        assert.throws(() => {
            const privKey = Buffer.from(PRIVATE_KEY, 'hex');
            return new Wallet(privKey, MNEMONIC);
        });
    });
});

describe('walletFromMnemonic()', () => {
    it('should throw on invalid mnemonic', () => {
        assert.throws(() => {
            walletFromMnemonic('a b c d e f g h i j k l');
        });
    });

    it('async', async () => {
        const wallet = walletFromMnemonic(MNEMONIC);
        const walletAsync = await walletFromMnemonic(MNEMONIC);
        assert.strictEqual(wallet.getPrivateKeyString(), walletAsync.getPrivateKeyString());
    });
});

describe('isValidMnemonic()', () => {
    it('should work on valid 15 words mnemonic', () => {
        assert.strictEqual(isValidMnemonic(MNEMONIC), true);
    });
    it('should work on valid 15 words mnemonic', () => {
        assert.strictEqual(isValidMnemonic(bip39.generateMnemonic(160)), true);
    });
    it('should work on valid 12 words mnemonic', () => {
        assert.strictEqual(isValidMnemonic(bip39.generateMnemonic()), true);
    });
    it('should fail on invalid 12 words mnemonic', () => {
        assert.strictEqual(isValidMnemonic('a b c d e f g h i j k l'), false);
    });
});


src/index.js
// @TODO remove some wordlists
import * as bip39 from 'bip39';
import { HDKey as hdKey } from 'ethereum-cryptography/hdkey';
import { isValidPrivate, privateToPublic, publicToAddress } from 'ethereumjs-util/dist/account';
import { publicToString } from 'minterjs-util';

function assert(val, msg) {
    if (!val) {
        throw new Error(msg || 'Assertion failed');
    }
}

/**
 * BIP39 Master seed from mnemonic phrase
 * @param mnemonic - 12 words
 * @return {Buffer}
 */
export function seedFromMnemonic(mnemonic) {
    return bip39.mnemonicToSeedSync(mnemonic);
}

/**
 * BIP39 Master seed from mnemonic phrase (async)
 * @param mnemonic - 12 words
 * @return {Promise<Buffer>}
 */
export function seedFromMnemonicAsync(mnemonic) {
    return bip39.mnemonicToSeed(mnemonic);
}

/**
 * BIP44 HD key from master seed
 * @param {Buffer} seed - 64 bytes
 * @return {HDKey}
 */
export function hdKeyFromSeed(seed) {
    // same as "m/44'/60'/0'/0/0"
    return hdKey.fromMasterSeed(seed).derive("m/44'/60'/0'/0").deriveChild(0);
}

/**
 * @param {Buffer} [priv]
 * @param {string} [mnemonic]
 * @param {boolean} [doAsync]
 * @constructor
 */
const Wallet = function WalletConstructor(priv, mnemonic, doAsync) {
    if (priv && mnemonic) {
        throw new Error('Cannot supply both a private and a mnemonic phrase to the constructor');
    }

    if (priv && !isValidPrivate(priv)) {
        throw new Error('Private key does not satisfy the curve requirements (ie. it is invalid)');
    }

    if (mnemonic && !bip39.validateMnemonic(mnemonic)) {
        throw new Error('Invalid mnemonic phrase');
    }

    if (mnemonic && doAsync) {
        return seedFromMnemonicAsync(mnemonic)
            .then((seed) => {
                this._privKey = hdKeyFromSeed(seed)._privateKey;
                this._mnemonic = mnemonic;

                return this;
            });
    }

    if (mnemonic) {
        const seed = seedFromMnemonic(mnemonic);
        priv = hdKeyFromSeed(seed)._privateKey;
    }

    this._privKey = priv;
    this._mnemonic = mnemonic;
};

Object.defineProperty(Wallet.prototype, 'mnemonic', {
    get() {
        assert(this._mnemonic, 'This is a private key only wallet');
        return this._mnemonic;
    },
});

Object.defineProperty(Wallet.prototype, 'privKey', {
    get() {
        return this._privKey;
    },
});

// uncompressed public key
Object.defineProperty(Wallet.prototype, 'pubKey', {
    get() {
        if (!this._pubKey) {
            this._pubKey = privateToPublic(this.privKey);
        }
        return this._pubKey;
    },
});

/**
 * @return {string}
 */
Wallet.prototype.getMnemonic = function getMnemonic() {
    return this.mnemonic;
};

/**
 * @return {Buffer}
 */
Wallet.prototype.getPrivateKey = function getPrivateKey() {
    return this.privKey;
};

/**
 * @return {string}
 */
Wallet.prototype.getPrivateKeyString = function getPrivateKeyString() {
    return `0x${this.getPrivateKey().toString('hex')}`;
};

/**
 * @return {Buffer}
 */
Wallet.prototype.getPublicKey = function getPublicKey() {
    return this.pubKey;
};

/**
 * @return {string}
 */
Wallet.prototype.getPublicKeyString = function getPublicKeyString() {
    return publicToString(this.getPublicKey());
};

/**
 * @return {Buffer}
 */
Wallet.prototype.getAddress = function getAddress() {
    return publicToAddress(this.pubKey);
};

/**
 * @return {string}
 */
Wallet.prototype.getAddressString = function getAddressString() {
    return `Mx${this.getAddress().toString('hex')}`;
};

/**
 * Generate Wallet from random mnemonic
 * @return {Wallet}
 */
export function generateWallet() {
    const mnemonic = bip39.generateMnemonic();
    return walletFromMnemonic(mnemonic);
}

/**
 * MinterWallet from mnemonic phrase
 * @param {string} mnemonic - 12 words
 * @return {Wallet}
 */
export function walletFromMnemonic(mnemonic) {
    return new Wallet(null, mnemonic);
}

/**
 * MinterWallet from mnemonic phrase
 * @param {string} mnemonic - 12 words
 * @return {Promise<Wallet>}
 */
export function walletFromMnemonicAsync(mnemonic) {
    return new Wallet(null, mnemonic, true);
}

/**
 * MinterWallet from private key
 * @param {Buffer} priv - 64 bytes
 * @return {Wallet}
 */
export function walletFromPrivateKey(priv) {
    return new Wallet(priv);
}

/**
 * Generate 12 words mnemonic phrase
 * @return {string}
 */
export function generateMnemonic() {
    return bip39.generateMnemonic();
}

/**
 * Check that mnemonic phrase has 12 words and represents valid entropy
 * @param {string} mnemonic
 * @return {boolean}
 */
export function isValidMnemonic(mnemonic) {
    return typeof mnemonic === 'string' && mnemonic.trim().split(/\s+/g).length >= 12 && bip39.validateMnemonic(mnemonic);
}

export default Wallet;
